llm/src/speed_monitor_w_mfu.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Monitor throughput during training."""
+from __future__ import annotations
+
+from collections import deque
+from typing import Any, Deque, Dict
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+from composer.utils import dist
+
+from composer.callbacks import SpeedMonitor
+
+GPU_AVAILABLE_FLOPS = 312_000_000_000_000
```

Comment:
Hmmm I think we should probably have this be a kwarg to `SpeedMonitorMFU`, or better yet, have it be a type like `Union[float, str]` and you can type in `h100`, `a100`, `v100`, `t4`, etc. I think we're a bit lucky to have A100s but some users of this repo may be on alternative hardware!

Reply:
Future: we probably want Composer to autodetect and expose the device info.. seems pretty useful. WandB figured how to do it:
<img width="700" alt="Screen Shot 2022-12-29 at 9 36 39 PM" src="https://user-images.githubusercontent.com/77638579/210028883-48e517e1-0948-4c6e-9d10-be3d84752d5f.png">

Reply:
`torch.cuda.get_device_name()` outputs something like: `'NVIDIA A100-SXM4-40GB'`
using this and hardcoding these gpu type flop counts:
```
GPU_AVAILABLE_FLOPS = {
    # source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
    'h100-sxm':     1.979e15 / 2,   # nvidia publishes spec sheet with a 2x sparsity factor
    'h100-pcie':    1.513e15 / 2,   # nvidia publishes spec sheet with a 2x sparsity factor
    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-us-nvidia-1758950-r4-web.pdf
    'a100':         312e12,         # sxm and pcie have same flop counts
    # source: https://images.nvidia.com/content/technologies/volta/pdf/tesla-volta-v100-datasheet-letter-fnl-web.pdf
    'v100-sxm':     125e12,
    'v100-pcie':    112e12,
    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/tesla-t4/t4-tensor-core-datasheet-951643.pdf
    't4':           65e12,          # sxm and pcie have same flop counts
}
```
Future: create something like a composer.gpu.constants tool for going thins

Reply:
Sorry to be pedantic, but I think we'll also need to hardcode these values per-precision type, and then look at the `state.precision` to figure out which value to use.

Reply:
updted to
```
GPU_AVAILABLE_FLOPS = {
    # source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
    # nvidia publishes spec sheet with a 2x sparsity factor
    'h100-sxm':     {'fp64': 67e12, 'fp32': 67e12, 'tf32': 989e12 / 2, 'fp16': 1.979e15 / 2, 'amp_fp16': 1.979e15 / 2, 'bf16': 1.979e15 / 2, 'amp_bf16': 1.979e15 / 2, 'fp8': 3_958 / 2, 'int8': 3_958 / 2},
    'h100-pcie':    {'fp64': 51e12, 'fp32': 51e12, 'tf32': 756e12 / 2, 'fp16': 1.513e15 / 2, 'amp_fp16': 1.513e15 / 2, 'bf16': 1.513e15 / 2, 'amp_bf16': 1.513e15 / 2, 'fp8': 3_026 / 2, 'int8': 3_026 / 2},
    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-us-nvidia-1758950-r4-web.pdf
    # sxm and pcie have same flop counts
    'a100':         {'fp64': 19.5e12, 'fp32': 19.5e12, 'tf32': 156e12, 'fp16': 312e12, 'amp_fp16': 312e12, 'bf16': 312e12, 'amp_bf16': 312e12},
    # source: https://images.nvidia.com/content/technologies/volta/pdf/volta-v100-datasheet-update-us-1165301-r5.pdf
    'v100-sxm':     {'fp64': 7.8e12, 'fp32': 15.7e12, 'fp16': 125e12, 'amp_fp16': 125e12},
    'v100-pcie':    {'fp64': 7e12, 'fp32': 14e12, 'fp16': 112e12, 'amp_fp16': 112e12},
    'v100s-pcie':   {'fp64': 8.2e12, 'fp32': 16.4e12, 'fp16': 130e12, 'amp_fp16': 130e12},
    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/tesla-t4/t4-tensor-core-datasheet-951643.pdf
    # sxm and pcie have same flop counts
    't4':           {'fp32': 8.1e12, 'fp16': 65e12, 'amp_fp16': 65e12, 'int8': 130e12, 'int4': 260e12},
}
```

its a bit verbose but works. lmk if you had something else in mind.